src/engine/AmguConsumer.sol

```diff
@@ -13,61 +13,77 @@ abstract contract AmguConsumer is DSMath {
     event AmguPaid(
         address indexed payer,
         uint256 totalAmguPaidInEth,
-        uint256 amguChargableGas,
-        uint256 incentivePaid
+        uint256 amguChargableGas
+    );
+
+    event IncentivePaid(
+        address indexed payer,
+        uint256 incentiveAmount
     );
 
     IRegistry public REGISTRY;
 
-    constructor (address _registry) public {
+    constructor(address _registry) public {
         REGISTRY = IRegistry(_registry);
     }
 
-    /// @dev if amgu price is zero, skip price fetching
-    /// @param _incentiveAmount Wei amount to be paid above AMGU
-    modifier amguPayableWithIncentive(uint256 _incentiveAmount) {
+    modifier amguPayable() {
+        uint256 preGas = gasleft();
+        _;
+        uint256 postGas = gasleft();
+        uint256 ethChargedForAmgu = __chargeAmgu(sub(preGas, postGas));
+        __refundExtraEther(ethChargedForAmgu);
+    }
+
+    modifier amguPayableWithIncentive() {
+        uint256 incentiveAmount = REGISTRY.incentive();
         require(
-            msg.value >= _incentiveAmount,
+            msg.value >= incentiveAmount,
             "amguPayableWithIncentive: Insufficent value for incentive"
         );
         uint256 preGas = gasleft();
         _;
         uint256 postGas = gasleft();
+        uint256 ethChargedForAmgu = __chargeAmgu(sub(preGas, postGas));
+        require(
+            msg.value >= incentiveAmount,
```

Comment:
`msg.value` is the original value of the msg, so need to compare it with the total amount charged, so maybe add a `uint256 totalEthCharged = add(ethChargedForAmgu, incentiveAmount)`, then use it here and in line 52